packages/loader/container-loader/src/serializedStateManager.ts

```diff
@@ -152,7 +159,19 @@ export class SerializedStateManager {
 				this.updateSnapshotAndProcessedOpsMaybe();
 			})();
 
-			return { baseSnapshot, version: undefined };
+			const blobContents = new Map<string, ArrayBuffer>();
+			for (const [id, value] of Object.entries(snapshotBlobs)) {
+				blobContents.set(id, stringToBuffer(value, "utf8"));
+			}
+			const iSnapshot: ISnapshot = {
+				sequenceNumber: this.snapshot.snapshotSequenceNumber,
+				snapshotTree: baseSnapshot,
+				blobContents,
+				latestSequenceNumber: undefined,
+				ops: [],
+				snapshotFormatV: 1,
+			};
+			return { baseSnapshot: iSnapshot, version: undefined };
 		}
```

Comment:
Is there a cheaper way of doing this? We don't always have to return ISnapshot here right? I'm assuming you have a reason for doing so.

Reply:
We pass the entire snapshot to container runtime if we have it. So, we need to return the entire snapshot to container from this, so that container can pass that to container runtime.

Reply:
@vladsud @tyler-cai-microsoft, what if we expose the snapshot blobs on the container context, so that runtime can access it as otherwise we do need to care about passing this entire thing to the runtime.

Reply:
Sounds good to me

Reply:
We will only deploy it once the saturation of appropriate FF package versions happens.

Reply:
Couple comments:
- our system should support back-compat, but it's purpose not to have back-compat forever. We want to actually deliver value faster, so if the right combination of bundles is on the machine, new functionality should shine.
    - So, the question is - if I have right version of driver / loader / runtime, will we go on new path here? I believe this requires this translation, so we need to it.
- We should stick to one format across layers for blobs. If some things use strings, other things use arrays, then we will be converting back and forth all the time. I think it's safe to assume that external serialization code (i.e. cache that consumes serialized container) cannot roundtrip arrays (as they are not serializable, i.e. calling JSON.stringify gives you garbage), so some code may need to stick with working with strings on the boundary.
    - That said, maybe we can learn here from ODSP driver caching. It uses arrays. And it works flawlessly (and efficiently) with indexDB. That said, some customers do not use indexDB, and they complained about arrays. I think the best course of action is for our APIs to be as efficient as possible (i.e. use arrays everywhere, if that's possible), but provide conversion tooling for people who need it.
    - In other words, make closeAndGetPendingLocalState() return unknown (or something else that says - I'm a tree and almost serializable, with exception of array leaf nodes), not a string.

Reply:
Adding @anthony-murphy to consider # 2 above. It's a perf optimization, so we do not need to consider it right away, but something to think about.

Reply:
Want to add something RE this question

> if we expose the snapshot blobs on the container context

I really dislike globals / singletons solutions (and IContainerContext is global for container runtime).
Instead, I'd prefer the following (this has nothing to do with string <-> array conversion discussion, but more on how things are exposed):
- ISnapshot is self-containing structure (as today), it's passed down and the layers below have all they have.
    - Though we might want a different interface for runtime to drop ops, as ops could be used only by loader layer.
- Whenever we need to pass data down the stack (for example to data store context / data store), we pass same ISnapshot type, but taking subtree, and recreating new ISnapshot out of ISnapshotTree by adding blobs property to it.
- This will over-expose blobs to layers down, but I think that's Ok.
- This way, every layer has everything they need to operate. There is no need to go back and ask container runtime for everything. For example, I think ContainerRuntime.isSnapshotFetchRequired should not exist - whatever questions data store layer wants to get answered it should be able to answer on its own.